### PR TITLE
Add getindex methods for NamedTuple

### PR DESCRIPTION
**Pull request by kescobo**
See #38834 

This adds `getindex()` methods for `NamedTuple`s to return multiple values as a new `NamedTuple`, eg:

```
julia> nt = (a=1, b=2, c=3)
(a = 1, b = 2, c = 3)

julia> nt[(:a, :b)]
(a = 1, b = 2)

julia> nt[[:a, :b]]
(a = 1, b = 2)

julia> nt[[:a,]]
(a = 1,)
```

To be analogous with `getindex` for `Tuple`s and `Vector`s. When Jeff commented on the issue, he only gave explicit approval to the idea of using a tuple of `Symbol`s to index, not the `AbstractVector`, but it's easy enough to remove the latter if unwanted. If the `AbstractVector` method is approved, in the tests, I wasn't sure if testing all of the methods in 1 test would be a good idea or not (see comments) - I assumed keeping them separate would be better.

I also added a bit to the docstring, but I wasn't sure if I should also add something to `base/docs/basedocs.jl` [somewhere around here](https://github.com/JuliaLang/julia/blob/master/base/docs/basedocs.jl#L2391-L2403)

**Open questions / todos**

- [x] OK to keep `getindex(t::NamedTuple, idxs::AbstractVector{Symbol})` ?
- [x] If yes, should tests for different methods be kept separate?
  - [x] either way, remove comments in file
- [x] Should I add something to `basedocs.jl`?
- [x] Should I add a compat notice to docstring like the [one here](https://github.com/JuliaLang/julia/blob/master/base/namedtuple.jl#L77-L78)? If yes, I assume it would be for 1.7 at this point?

If there's anything else I've missed, please let me know!

### CODE REVIEW

**Comment by kescobo** [comment]
More stupid typos :facepalm: - got it fixed locally but will wait to push so as not to trigger extra CI time

**Comment by kescobo** [comment]
Bump! Was wondering if maybe failing CI was keeping people from reviewing so I pushed the typo fix.

File: base/namedtuple.jl [hunk 1]
```diff
@@ -116,6 +119,8 @@ firstindex(t::NamedTuple) = 1
 lastindex(t::NamedTuple) = nfields(t)
 getindex(t::NamedTuple, i::Int) = getfield(t, i)
 getindex(t::NamedTuple, i::Symbol) = getfield(t, i)
+@inline getindex(t::NamedTuple, idxs::Tuple{Vararg{Symbol}}) = NamedTuple{idxs}(t)
```
Comment (quinnj) [inline base/namedtuple.jl:122]: Sorry if this has been discussed elsewhere, but why the `@inline`? Seems to not match the other methods around and should be simple enough inline by itself?

Comment (kescobo) [reply base/namedtuple.jl:122]: That's how Jeff wrote it in the attached issue (#38834) - TBH I still don't totally understand what `@inline` does :shrug: . Happy to leave or remove

File: test/namedtuple.jl [hunk 2]
```diff
@@ -21,6 +21,13 @@
 @test (a=3,)[:a] == 3
 @test (x=4, y=5, z=6).y == 5
 @test (x=4, y=5, z=6).z == 6
+@test (x=4, y=5, z=6)[(:x, :y)] == (x=4, y=5)
+@test (x=4, y=5, z=6)[(:x,)] == (x=4,)
+# If the following are included, is it worth putting in 1 line with above, eg:
+# @test (x=4, y=5, z=6)[(:x, :y)] == (x=4, y=5, z=6)[[:x, :y]] == (x=4, y=5)
+# or keep separate, eg:
+@test (x=4, y=5, z=6)[[:x, :y]] == (x=4, y=5)
+@test (x=4, y=5, z=6)[[:x]] == (x=4,)
```
Comment (jw3126) [inline test/namedtuple.jl:30]: Could add some `@test_throws` on stuff like `(a=1,)[(:x,)]` or `(x=1, y=2)[(:x,:x)]` also corner cases like `(a=...)[()]` and `NamedTuple()[()]` could be added.

Comment (kescobo) [reply test/namedtuple.jl:30]: Hmm - good thinking. For corner cases, what are the expected behaviors? For the later, I think I might expect it to return another empty tuple, since 

```
julia> [][:]
Any[]
```

But I don't know. And for the first of your corner cases, I'm assuming you mean `(a=...)` as shorthand for an actual named tuple? Or did you mean something using splatting?

Comment (jw3126) [reply test/namedtuple.jl:30]: Yes, I meant e.g. `@test (a=1, b=2)[()] === NamedTuple()` and in general `@test nt[()] === NamedTuple()`

**Comment by kescobo** [comment]
Happy New Year! Bump :-P

File: test/namedtuple.jl [hunk 3]
```diff
@@ -21,9 +21,23 @@
 @test (a=3,)[:a] == 3
 @test (x=4, y=5, z=6).y == 5
 @test (x=4, y=5, z=6).z == 6
+@test (x=4, y=5, z=6)[(:x, :y)] == (x=4, y=5)
+@test (x=4, y=5, z=6)[(:x,)] == (x=4,)
+# If the following are included, is it worth putting in 1 line with above, eg:
+# @test (x=4, y=5, z=6)[(:x, :y)] == (x=4, y=5, z=6)[[:x, :y]] == (x=4, y=5)
```
Comment (quinnj) [inline test/namedtuple.jl:27]: I think it's fine to leave the extra line out; if you remove these lines, I think this is ready to merge

**Comment by kescobo** [comment]
Alright, removed comments from tests, and also rebased onto current master branch. Open questions: 

1. should I addto `basedocs,jl`? I think not - looking again, other methods there are for constructors
2. Should I add `compat` notice? I think it would be good, but want to double check that it should be for 1.7

**Comment by quinnj** [comment]
Yeah, add a compat notice

**Comment by simeonschaub** [comment]
Probably also worth adding an entry in NEWS.md
